tempesta_fw/lb/tfw_lb_http.c

```diff
@@ -0,0 +1,1091 @@
+ */
+#define TFW_LB_GROUP_SIZE(sks_n) \
+	(sizeof(TfwLbGroup) * ((sks_n) + 1) * sizeof(struct sk *))
+
+/**
+ * XXX: the structure is adopted from tfw_sched_http. It allows to route HTTP
+ * requests to a certain group of backend servers depending on HTTP request
+ * fields: URI, Host, headers, etc.
+ */
+typedef struct {
+	TfwLbGroup *group;
+	TfwHttpMatchRule rule;
+} TfwLbRule;
+
+/**
+ * The list of all groups of backend servers (consists of TfwLbGroup objects).
+ */
+static LIST_HEAD(tfw_lb_groups);
+
+/**
+ * The list of rules that allow to choose a particular group for sending a
+ * HTTP request depending on its fields: URI, Host, headers. etc.
+ */
+static TfwHttpMatchList *tfw_lb_rules;
+
+/**
+ * The group which is chosen when none of tfw_lb_rules is applied.
+ */
+static TfwLbGroup *tfw_lb_default_group;
+
+/**
+ * The array of round-robin counters for tfw_lb_groups.
+ * Each TfLbGroup has a corresponding slot allocated in this array.
+ * The slot may be accessed via TFW_LB_GROUP_RR_CTR(TfwLbGroup->rr_idx).
+ *
+ * XXX: non-shared round-robin counters allow to avoid atomic instructions
+ * and prevent cache line bouncing.
+ */
+static DEFINE_PER_CPU(u8[TFW_LB_GROUP_MAX], tfw_lb_group_rr_ctrs);
+
+#define TFW_LB_GROUP_RR_CTR(group) \
+	(__get_cpu_var(tfw_lb_group_rr_ctrs)[(group)->rr_idx])
+
+/**
+ * The number entries in the tfw_lb_groups list.
+ * Helps to protect tfw_lb_group_rr_ctrs from overflow.
+ */
+static int tfw_lb_groups_n;
+
+/*
+ * ------------------------------------------------------------------------
+ *	tfw_lb_http_send_msg() and its scheduling helpers
+ * ------------------------------------------------------------------------
+ *
+ * This is the actual load balancing implementation.
+ * The code decides how to distribute HTTP requests across backend servers.
+ */
+
+static struct sock *
+tfw_lb_sched_rr_failover(TfwLbGroup *group)
+{
+	int sks_n, ctr, ctr_end;
+	struct sock *sk, **sks;
+
+	sks = group->sks;
+	sks_n = group->sks_n;
+	ctr = TFW_LB_GROUP_RR_CTR(group);
+	ctr_end = ctr + sks_n;
+
+	while (likely(++ctr < ctr_end)) {
+		sk = sks[ctr % sks_n];
+		if (likely(sk->sk_state == TCP_ESTABLISHED)) {
+			TFW_LB_GROUP_RR_CTR(group) = ctr;
+			return sk;
+		}
+	}
+
+	TFW_ERR("can't find any online backend server\n");
+	return NULL;
+}
+
+/**
+ * XXX: the whole tfw_sched_rr is collapsed into this single function.
+ */
+static struct sock *
+tfw_lb_sched_rr(TfwLbGroup *group)
+{
+	struct sock *sk;
+
+	sk = group->sks[++TFW_LB_GROUP_RR_CTR(group) % group->sks_n];
+	if (likely(sk->sk_state == TCP_ESTABLISHED))
+		return sk;
+
+	return tfw_lb_sched_rr_failover(group);
+}
+
+/**
+ * XXX: the whole tfw_sched_hash is collapsed into this single function.
+ * Rendezvous hashing is included as a failovering strategy.
+ *
+ * XXX: the linear search is slow but it is required for failover switching.
+ * Perhaps we should do binary search first, and then fall back to linear if
+ * the chosen socket is dead.
```

Comment:
Well, basically even with https://github.com/natsys/tempesta/issues/76 in mind we expect a lot of groups, but not so many servers for one group among which we do load balancing and failovering, so linear scan over contigouse cache line aligned array is Ok.